render/render.c

```diff
@@ -0,0 +1,378 @@
+	mat[4] = y * t[3];
+
+	// Translation
+	mat[2] = -copysign(1.0f, mat[0] + mat[1]);
+	mat[5] = -copysign(1.0f, mat[3] + mat[4]);
+
+	// Identity
+	mat[8] = 1.0f;
+}
+
+void wlr_render_bind_raw(struct wlr_render *rend, uint32_t width, uint32_t height,
+		enum wl_output_transform transform) {
+	assert(eglGetCurrentContext() == rend->egl->context);
+	DEBUG_PUSH;
+
+	glViewport(0, 0, width, height);
+	matrix(rend->proj, width, height, transform);
+
+	DEBUG_POP;
+}
+
+void wlr_render_bind(struct wlr_render *rend, struct wlr_output *output) {
+	wlr_render_bind_raw(rend, output->width, output->height, output->transform);
+}
+
+void wlr_render_clear(struct wlr_render *rend, float r, float g, float b, float a) {
+	assert(eglGetCurrentContext() == rend->egl->context);
+	DEBUG_PUSH;
+
+	glClearColor(r, g, b, a);
+	glClear(GL_COLOR_BUFFER_BIT);
+
+	DEBUG_POP;
+}
+
+void wlr_render_subtexture(struct wlr_render *rend, struct wlr_tex *tex,
+		int32_t tex_x1, int32_t tex_y1, int32_t tex_x2, int32_t tex_y2,
+		int32_t pos_x1, int32_t pos_y1, int32_t pos_x2, int32_t pos_y2, int32_t pos_z) {
+	assert(eglGetCurrentContext() == rend->egl->context);
+	DEBUG_PUSH;
+
+	GLuint prog = rend->shaders.extn;
+
+	GLuint proj_loc = glGetUniformLocation(prog, "proj");
+	GLuint pos_loc = glGetAttribLocation(prog, "pos");
+	GLuint texcoord_loc = glGetAttribLocation(prog, "texcoord");
+
+	glUseProgram(rend->shaders.extn);
+	glUniformMatrix3fv(proj_loc, 1, GL_TRUE, rend->proj);
+
+	glActiveTexture(GL_TEXTURE0);
+	glBindTexture(GL_TEXTURE_EXTERNAL_OES, tex->image_tex);
+
+	GLfloat verts[] = {
+		pos_x1, pos_y1, pos_z,
+		pos_x2, pos_y1, pos_z,
+		pos_x1, pos_y2, pos_z,
+		pos_x2, pos_y2, pos_z,
+	};
+
+	GLfloat tw = tex->width;
+	GLfloat th = tex->height;
+	// Transform to OpenGL [0,1] texture coordinate space
+	GLfloat tx1 = tex_x1 / tw;
+	GLfloat tx2 = tex_x2 / tw;
+	GLfloat ty1 = tex_y1 / th;
+	GLfloat ty2 = tex_y2 / th;
+
+	GLfloat texcoord[] = {
+		tx1, ty1,
+		tx2, ty1,
+		tx1, ty2,
+		tx2, ty2,
+	};
+
+	glVertexAttribPointer(pos_loc, 3, GL_FLOAT, GL_FALSE, 0, verts);
+	glVertexAttribPointer(texcoord_loc, 2, GL_FLOAT, GL_FALSE, 0, texcoord);
+
+	glEnable(GL_BLEND);
+	glBlendFunc(GL_SRC_ALPHA, GL_ONE_MINUS_SRC_ALPHA);
+
+	glEnableVertexAttribArray(pos_loc);
+	glEnableVertexAttribArray(texcoord_loc);
+
+	glDrawArrays(GL_TRIANGLE_STRIP, 0, 4);
+
+	DEBUG_POP;
+}
+
+bool wlr_render_read_pixels(struct wlr_render *rend, enum wl_shm_format wl_fmt,
+		uint32_t stride, uint32_t width, uint32_t height,
+		uint32_t src_x, uint32_t src_y, uint32_t dst_x, uint32_t dst_y,
+		void *data) {
+	assert(eglGetCurrentContext() == rend->egl->context);
+
+	const struct format *fmt = wl_to_gl(wl_fmt);
+	if (!fmt) {
+		wlr_log(L_ERROR, "Unsupported pixel format");
+		return false;
+	}
+
+	DEBUG_PUSH;
+
+	// Make sure any pending drawing is finished before we try to read it
+	glFinish();
+
+	// Unfortunately GLES2 doesn't support GL_PACK_*, so we have to read
+	// the lines out row by row
```

Comment:
Is there a way to speed this up in the special case where `stride == 4*width`?

Reply:
We're actually reading though it backwards as well, so probably not.

Reply:
That's true. If we have issues with performance (e.g. trying to get a video stream with the screenshooter interface) we could make only one call to `glReadPixels` and flip it, that may be faster than `height` calls to `glReadPixels`.

Reply:
If this was supposed to be high performance, I would make it hand dma bufs around, instead of raw pixels.

Reply:
This is probably why we're going to want a GLES3 renderer.